File: Tests/original-code/matlab-code/main_SHO.m
%_______________________________________________________________________________________%
%  Sea-Horse optimizer (SHO)
%  Developed in MATLAB R2018a
%
%  programmer: Shijie Zhao and Tianran Zhang   
%  E-mail: zhaoshijie@lntu.edu.cn
%          ztr20010118@126.com
%  The code is based on the following papers.
%  Shijie Zhao, Tianran Zhang, Shilin Ma, Mengchen Wang 
%  Sea-horse optimizer: a novel nature-inspired meta-heuristic for global
%  optimization problems.
%  Applied Intelligence
%_______________________________________________________________________________________%
clear all 
clc

popsize=30; % Number of search agents
Max_iter=500; % Maximum iteration
F_name='F1'; % Name of the test function that can be from F1 to F23 (Table 2,3,4 in the paper)

[LB,UB,Dim,fobj]=BenchmarkFunctions(F_name);% Load details of the selected benchmark function

tic
[ObjectiveFitness,ObjectivePosition,Convergence_curve,Trajectories,fitness_history, population_history]=SHO(popsize,Max_iter,LB,UB,Dim,fobj);
time=toc;

semilogy(1:Max_iter,Convergence_curve,'color','r','linewidth',2.5);
title('Convergence curve');
xlabel('Iteration');
ylabel('Best score obtained so far')
display(['The running time is:', num2str(time)]);
display(['The best solution obtained by SHO is : ', num2str(ObjectiveFitness)]);
display(['The best optimal sea horse of the objective funciton found by SHO is : ', num2str(ObjectivePosition)]);

File: Tests/original-code/python-code/main_SHO.py
"""Main entry point for running the Sea-Horse Optimizer benchmark demo."""

from __future__ import annotations

import time

import matplotlib.pyplot as plt
import numpy as np

from BenchmarkFunctions import BenchmarkFunctions
from SHO import SHO


def main() -> None:
    popsize = 30  # Number of search agents
    max_iter = 500  # Maximum iteration
    f_name = "F1"  # Name of the benchmark function (F1 to F23)

    lb, ub, dim, fobj = BenchmarkFunctions(f_name)

    start = time.time()
    objective_fitness, objective_position, convergence_curve, trajectories, fitness_history, population_history = SHO(
        popsize, max_iter, lb, ub, dim, fobj
    )
    elapsed = time.time() - start

    plt.semilogy(np.arange(1, max_iter + 1), convergence_curve, color="r", linewidth=2.5)
    plt.title("Convergence curve")
    plt.xlabel("Iteration")
    plt.ylabel("Best score obtained so far")
    plt.show()

    print(f"The running time is: {elapsed}")
    print(f"The best solution obtained by SHO is : {objective_fitness}")
    print(f"The best optimal sea horse of the objective function found by SHO is : {objective_position}")

    # Keep variables available for parity with MATLAB main script behavior.
    _ = trajectories, fitness_history, population_history


if __name__ == "__main__":
    main()
